libscript/include/pivot/script/Parser.hxx

```diff
@@ -0,0 +1,89 @@
+#pragma once
+
+#include "pivot/script/DataStructures.hxx"
+
+#include <queue>
+
+namespace pivot::ecs::script::parser
+{
+
+/*! \brief	Parser/Lexer handler to parse files and extract ASTs
+ *
+ * 	The lexer will analyze files and extract tokens based on symbols.
+ * 	It will pass these tokens to the parser which will extract nodes
+ * 	and generate an AST based on syntax.
+ *
+ */
+class Parser
+{
+public:
+    ///	Main function, will return the abstract syntax tree of the file (root Node)
+    Node ast_from_file(const std::string &file, bool isContent = false, bool verbose = true);
+    /// Lexer function, will return a list of Tokens contained in the file
+    void tokens_from_file(const std::string &file, bool isContent = false, bool verbose = false);
```

Comment:
Take an istream and remove `isContent`
```suggestion
    ///	Main function, will return the abstract syntax tree of the file (root Node)
    Node ast_from_file(std::istream input, bool verbose = true);
    /// Lexer function, will return a list of Tokens contained in the file
    void tokens_from_file(std::istream input, bool verbose = false);
```

Reply:
This would allow using a single parameter which could act both as a file, or a harcoded string, which is exactly what I was looking for.
However, I don't understand how to differentiate that once my function has been called.
How can i know the file name (to display to the user in case of error for example) or distinguish between file/string if the stream has been constructed before the call to the function ?

Reply:
For giving context to the user, you could add the file name as a parameter, with some default value.
To read from the stream, you don't need to know whether the underlying source is a string, a file, or anything else. You never need to distinguish between the two.